tests/bgp/test_startup_tsa_tsb_service.py

```diff
@@ -0,0 +1,725 @@
+        res = suphost.shell(cmd)
+        if not res.is_successful:
+            pytest.fail('DUT {} run command {} failed'.format(suphost.hostname, cmd))
+
+        # Waiting for SSH connection shutdown
+        pytest_assert(check_ssh_state(localhost, sup_ip, SSH_STATE_ABSENT, SSH_SHUTDOWN_TIMEOUT),
+                      'DUT {} did not shutdown'.format(suphost.hostname))
+        # Waiting for SSH connection startup
+        pytest_assert(check_ssh_state(localhost, sup_ip, SSH_STATE_STARTED, SSH_STARTUP_TIMEOUT),
+                      'DUT {} did not startup'.format(suphost.hostname))
+
+        logging.info("Wait until all critical processes are fully started")
+        wait_critical_processes(suphost)
+
+        sup_uptime = suphost.get_up_time()
+        logger.info('DUT {} up since {}'.format(suphost.hostname, sup_uptime))
+        rebooted = float(sup_uptime_before.strftime("%s")) != float(sup_uptime.strftime("%s"))
+        assert rebooted, "Device {} did not reboot".format(suphost.hostname)
+
+        for linecard in duthosts.frontend_nodes:
+            wait_for_startup(linecard, localhost, delay=10, timeout=300)
+
+            # Ensure startup_tsa_tsb service started on expected time since dut rebooted
+            dut_uptime = linecard.get_up_time()
+            logging.info('DUT {} up since {}'.format(linecard.hostname, dut_uptime))
+            service_uptime = get_tsa_tsb_service_uptime(linecard)
+            time_diff = (service_uptime - dut_uptime).total_seconds()
+            pytest_assert(int(time_diff) < 120,
+                          "startup_tsa_tsb service started much later than the expected time after dut reboot")
+
+            # Verify DUT is in maintenance state.
+            pytest_assert(TS_MAINTENANCE == get_traffic_shift_state(linecard),
+                          "DUT is not in maintenance state when startup_tsa_tsb service is running")
+
+            logging.info("Wait until all critical processes are fully started")
+            wait_critical_processes(linecard)
+            pytest_assert(wait_until(600, 20, 0, check_interface_status_of_up_ports, linecard),
+                          "Not all ports that are admin up on are operationally up")
+
+            pytest_assert(verify_only_loopback_routes_are_announced_to_neighs(
+                duthosts, linecard, dut_nbrhosts[linecard], traffic_shift_community),
+                "Failed to verify routes on nbr in TSA")
+
+        # Once all line cards are in maintenance state, proceed further
+        for linecard in duthosts.frontend_nodes:
+            # Verify startup_tsa_tsb service stopped after expected time
+            pytest_assert(wait_until(tsa_tsb_timer[linecard], 20, 0, get_tsa_tsb_service_status, linecard, 'exited'),
+                          "startup_tsa_tsb service is not stopped even after configured timer expiry")
+
+            pytest_assert(wait_until(90, 5, 0, check_tsc_command_error, linecard),
```

Comment:
why is this required? In what cases does TSC error out after TSA-TSB timer expiry?

Reply:
Not required actually. Added it for debugging purposes initially. Removed it now.